src/main/scala/scala/compat/java8/collectionImpl/Stepper.scala

```diff
@@ -0,0 +1,545 @@
+    * This is a terminal operation, though if `knownSize` is non-negative, it won't actually
+    * iterate over the elements.
+    */
+  def count(): Long = knownSize match {
+    case x if x < 0 => var n = 0L; while (hasStep) { nextStep; n += 1 }; n
+    case x => x
+  }
+
+  /** Consumes all remaining elements in this `Stepper` and counts how many satisfy condition `p`.
+    * This is a terminal operation.
+    */
+  def count(p: A => Boolean): Long = { var n = 0L; while (hasStep) { if (p(nextStep)) n += 1 }; n }
+
+  /** Searches for an element that satisfies condition `p`.  If none are found, it returns `false`.
+    * This is a terminal operation.
+    */
+  def exists(p: A => Boolean): Boolean = { while(hasStep) { if (p(nextStep)) return true }; false }
+
+  /** Searches for an element that satisifes condition `p`, returning it wrapped in `Some` if one is found, or `None` otherwise.
+    * This is a terminal operation.
+    */
+  def find(p: A => Boolean): Option[A] = { while (hasStep) { val a = nextStep; if (p(a)) return Some(a) }; None }
+
+  /** Repeatedly applies `op` to propagate an initial value `zero` through all elements of the collection.
+    * Traversal order is left-to-right.
+    * This is a terminal operation.
+    */
+  def fold[@specialized(Double, Int, Long) B](zero: B)(op: (B, A) => B) = { var b = zero; while (hasStep) { b = op(b, nextStep) }; b }
+
+  /** Repeatedly applies `op` to propagate an initial value `zero` through the collection until a condition `p` is met.
+    * If `p` is never met, the result of the last operation is returned.
+    * This is a terminal operation.
+    */
+  def foldTo[@specialized(Double, Int, Long) B](zero: B)(op: (B, A) => B)(p: B => Boolean) = { var b = zero; while (!p(b) && hasStep) { b = op(b, nextStep) }; b }
+
+  /** Applies `f` to every remaining element in the collection.
+    * This is a terminal operation.
+    */
+  def foreach(f: A => Unit) { while (hasStep) f(nextStep) }
+
+  /** Repeatedly merges elements with `op` until only a single element remains.
+    * Throws an exception if the `Stepper` is empty.
+    * Merging occurs from left to right.
+    * This is a terminal operation.
+    */
+  def reduce(op: (A, A) => A): A = { var a = nextStep; while (hasStep) { a = op(a, nextStep) }; a }
+
+
+  ////
+  // Operations that convert to another related type
+  ////
+
+  /** Returns this `Stepper` as a `java.util.Spliterator`.
+    * This is a terminal operation.
+    */
+  def spliterator: Spliterator[A]
+
+  /** Returns this `Stepper` as a Scala `Iterator`.
+    * This is a terminal operation.
+    */
+  def iterator: Iterator[A] = new scala.collection.AbstractIterator[A] {
+    def hasNext = self.hasStep
+    def next = self.nextStep
+  }
+
+  /** Returns a Scala collection of the type requested. */
+  def to[Coll[_]](implicit cbf: collection.generic.CanBuildFrom[Nothing, A, Coll[A]]): Coll[A] = {
+    val b = cbf()
+    while (hasStep) b += nextStep
```

Comment:
It's a tradeoff.  `foreach` has an extra closure, which makes the optimization harder in the case where it's just doing this.  Since in my implementations it almost always does just this, I went with this one.  I don't think I benchmarked extensively enough to know what is better and worse by how much, though.